geps/gep-1731/index.md

```diff
@@ -0,0 +1,426 @@
+    //
+    // An implementation MAY use an exponential or alternative backoff strategy
+    // for subsequent retry attempts, MAY cap the maximum backoff duration to
+    // some amount greater than the specified minimum, and MAY add arbitrary
+    // jitter to stagger requests, as long as unsuccessful backend requests are
+    // not retried before the configured minimum duration.
+    //
+    // If a Request timeout (`rules[].timeouts.request`) is configured on the
+    // route, the entire duration of the initial request and any retry attempts
+    // MUST not exceed the Request timeout duration. If any retry attempts are
+    // still in progress when the Request timeout duration has been reached,
+    // these SHOULD be cancelled if possible and the Gateway MUST immediately
+    // return a timeout error.
+    //
+    // If a BackendRequest timeout (`rules[].timeouts.backendRequest`) is
+    // configured on the route, any retry attempts which reach the configured
+    // BackendRequest timeout duration without a response SHOULD be cancelled if
+    // possible and the Gateway should wait for at least the specified backoff
+    // duration before attempting to retry the backend request again.
+    //
+    // If a BackendRequest timeout is _not_ configured on the route, retry
+    // attempts MAY time out after an implementation default duration, or MAY
+    // remain pending until a configured Request timeout or implementation
+    // default duration for total request time is reached.
+    //
+    // When this field is unspecified, the time to wait between retry attempts
+    // is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Backoff *Duration `json:"backoff,omitempty"`
+}
+
+// HTTPRouteRetryStatusCode defines an HTTP response status code for
+// which a backend request should be retried.
+//
+// Implementations MUST support the following status codes as retryable:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// Implementations SHOULD support the `"5xx"` shorthand for matching status
+// codes in the 500-599 range. For implementations supporting `"5xx"`, this
+// shorthand MUST match the following status codes:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// The `"5xx"` shorthand MAY additionally match a well-documented subset of
+// arbitrary status codes in the 500-599 range, or match all status codes in
+// this range.
+//
+// Implementations SHOULD NOT support retrying status codes in the 100-399
+// range, as these responses are generally not appropriate to retry.
+//
+// Implementations MAY support specifying discrete values in the 400-499 range,
+// which are often inadvisable to retry, and MAY support discrete values in the
+// 600-999 (inclusive) range, which are not valid for HTTP clients, but are
+// sometimes used for communicating application-specific errors.
+//
+// Implementations MAY support additional shorthand codes for any `[0-9]xx`
+// range.
+//
```

Comment:
```suggestion
```
Is this a bad idea? Open to dropping this.

Reply:
I think we should probably actually follow the pattern we've used elsewhere, as annoying as it is:

1. Only the values defined by Gateway API do not need a prefix (any int status code within a specified range)
2. Everything else needs to be domain-prefixed until it's defined by Gateway API

That's admittedly rather annoying, but it does ensure that if/when we want to define 5xx or other similar shortcuts, we can without trampling over existing uses.

Reply:
Hmmm, feels like that would be slightly awkward for an end user, but does at least clearly indicate they're using non-standard functionality. TBH I can't forsee any range beyond 5xx having a path to becoming standardized given the stability of the upstream HTTP status code specification here, but okay adding this formatting.

Reply:
Hmmm, feels like that would be slightly awkward for an end user, but does at least clearly indicate they're using non-standard functionality. TBH I can't forsee any range beyond 5xx having a path to becoming standardized given the stability of the upstream HTTP status code specification here, but okay either adding this formatting or dropping this suggestion entirely.

```suggestion
// Implementations MAY support additional shorthand codes for any `[0-9]xx`
// range, but if doing so MUST use an implementation-specific domain prefix,
// such as `envoyproxy.io/4xx`.
//
```

Reply:
Thinking about this more, I'd really rather remove this line, as it feels like it will immediately be stretched for implementation-specific "retry on" semantics beyond status codes to mirror existing proxy configuration (e.g. `envoyproxy.io/envoy-ratelimited`), and allowing this implicitly guides users toward non-portable configuration.